csrc/optimization/move_split_cat.cpp

```diff
@@ -0,0 +1,140 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <optimization/move_split_cat.h>
+
+#include <vector>
+
+#include <expr_simplifier.h>
+#include <fusion.h>
+#include <ir/builder.h>
+#include <ir/interface_nodes.h>
+#include <ir/internal_base_nodes.h>
+#include <ir/utils.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+// Finds the canceling split of `cat` and returns the input TensorView of the
+// split. A split (implemented as multiple `slice`s) and a cat cancel when they
+// work on the same dimension. For example, when
+//
+//   s0 = in[:, :5]
+//   s1 = in[:, 5:]
+//   out = cat([s0, s1], dim=-1)
+//
+// findCancelingSplit(out) returns `in`.
+//
+// Currently, the cat has to immediately follow the split. But we will soon
+// handle patterns like `split ->
+// a_chain_of_unary_ops_that_do_not_resize_the_split_dimension -> cat`.
+TensorView* findCancelingSplit(CatOp* cat) {
+  // nvFuser implements cat as PadOp->CatOp. We first locate these `PadOp`s.
+  std::vector<PadOp*> pads;
+  pads.reserve(cat->inputs().size());
+  for (Val* in : cat->inputs()) {
+    pads.push_back(in->definition()->as<PadOp>());
+  }
+
+  // We then locate the `SliceOp`s right before these `PadOp`s.
+  std::vector<SliceOp*> slices;
+  slices.reserve(pads.size());
+  for (PadOp* pad : pads) {
+    auto* slice = dynamic_cast<SliceOp*>(pad->in()->definition());
+    if (slice == nullptr) {
+      return nullptr;
+    }
+    slices.push_back(slice);
+  }
+
+  // Check the slices are based on the same tensor. Otherwise, they don't form a
+  // split.
+  TensorView* split_in = nullptr;
+  for (SliceOp* slice : slices) {
+    if (split_in == nullptr) {
+      split_in = slice->in();
+    } else if (split_in != slice->in()) {
+      return nullptr;
+    }
+  }
+  NVF_ERROR(split_in != nullptr);
```

Comment:
should we error out here, or should we just return nullptr?

We can legitimately concat two sliced tensors.

Reply:
> We can legitimately concat two sliced tensors.

Yes, and that's checked by the `return nullptr` at line 62. 

This `NVF_ERROR` is for `cat`s that have no inputs. Is that even possible? Maybe `NVF_CHECK` is more appropriate.

Reply:
> Yes, and that's checked by the return nullptr at line 62.

How blind was I :face_with_head_bandage: